### ISSUE TITLE
Dungeon generator: Stairs frequently go one node too deep

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#8865]
##### Minetest version
```
5.0.1
```

##### Summary
The stairs of dungeons often go one node too far downwards.

At the bottom end of the stairs, there should be a normal floor, but instead there's an additional stair that oddly digs into the floor of the room. This looks odd.

This affects both “classic” cobblestone dungeons and the large desert stone (and similar) dungeons.

This bug is very common, but it does not occur with 100% probability. Some stairways are still correct.

Screenshot (taken in v7 mapgen):
https://satoshiupload.com/images/OCB8v7P441.png

Expected behaviour: The excess stair should be a normal full cube instead, then it would be correct.

##### Steps to reproduce
* Create map in mapgen v7 and seed = 0
* Teleport to coordinates shown on screenshot

**Comment by paramat** [comment]
Yes, i've noticed this. Dungeon stair code is tricky, maybe i'll tackle it again sometime.

**Comment by paramat** [comment]
![screenshot_20190909_004057](https://user-images.githubusercontent.com/3686677/64553063-926cb700-d330-11e9-9245-bbcf3de364fc.png)

Altering code to have corridors of width 3 nodes, it gets worse, with a trench at the top of the stairs and gaps under them. I think stair code was only tested for up to 2 node wide corridors.
